src/utils/augmentation.py

```diff
@@ -0,0 +1,152 @@
+                symbols.append(random_symb)
+            else:
+                symbols.append(ch)
+        return "".join(symbols)
+
+
+class DiacriticsAugmentation(AugmentationWord):
+    """
+    Adds diacritics to chars of the word with a given probability
+    """
+
+    def __init__(self, expected_number_of_chars: float, count_adds: int = 1):
+        self.expected_number_of_chars = expected_number_of_chars
+        self.count_adds = count_adds
+
+    def __call__(self, word: str) -> str:
+        symbols = []
+        proba_per_char = (
+            self.expected_number_of_chars / len(word) if len(word) >= self.expected_number_of_chars else 0.5
+        )
+        for ch in word:
+            replace = np.random.binomial(1, proba_per_char)
+            if replace:
+                char_with_diac = ch
+                for i in range(self.count_adds):
+                    randBytes = random.randint(0x300, 0x36F).to_bytes(2, "big")
+                    char_with_diac += randBytes.decode("utf-16be")
+                symbols.append(char_with_diac)
+            else:
+                symbols.append(ch)
+        return "".join(symbols)
+
+
+class SpaceAugmentation(AugmentationWord):
+    """
+    Adds spaces to chars of the word with a given probability
+    """
+
+    def __init__(self, expected_number_of_chars: float):
+        self.expected_number_of_chars = expected_number_of_chars
+
+    def __call__(self, word: str) -> str:
+        symbols = []
+        proba_per_char = (
+            self.expected_number_of_chars / len(word) if len(word) >= self.expected_number_of_chars else 0.5
+        )
+        for ch in word:
+            replace = np.random.binomial(1, proba_per_char)
+            if replace:
+                symbols.append(" ")
+            symbols.append(ch)
+        return "".join(symbols)
+
+
+class AugmentationText:
+    """
+    Generates a noisy text with given parameters:
+    text - original text to which word augmentations are applied
+    proba_per_text - probability of noise for a given text;
+    EXPECTED_NUMBER_OF_WORDS - expected value (average) of words in every text that we want to make noisy
+    EXPECTED_NUMBER_OF_CHARS - expected value of chars in a word that we want to make noisy
+    MAX_COUNT_AUGM - maximum value of augmentations that can be applied to every word
+    """
+
+    def __init__(self, leet_symbols: dict, cluster_symbols: dict, proba_per_text: float):
+        diacritics = DiacriticsAugmentation(EXPECTED_NUMBER_OF_CHARS)
+        clusters = SimilarCharAugmentation(cluster_symbols, EXPECTED_NUMBER_OF_CHARS)
+        leet = SimilarCharAugmentation(leet_symbols, EXPECTED_NUMBER_OF_CHARS)
+        spaces = SpaceAugmentation(EXPECTED_NUMBER_OF_CHARS)
+        swap = SwapAugmentation()
+        self.augmentations = np.array([diacritics, clusters, leet, spaces, swap])
+        self.augmentations_probas = [0.3, 0.3, 0.3, 0.05, 0.05]  # sum must be equal to 1
```

Comment:
Why equal to 1?

Reply:
Because there is no auto-normalization in np.random.choice for probabilities (can cause _ValueError: probabilities do not sum to 1_)